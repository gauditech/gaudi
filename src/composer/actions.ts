import _ from "lodash";

import { FilteredByKind } from "@src/common/kindFilter";
import {
  assertUnreachable,
  ensureEmpty,
  ensureEqual,
  ensureExists,
  resolveItems,
} from "@src/common/utils";
import { getTypeModel } from "@src/compiler/ast/type";
import { composeValidators, validateFieldType } from "@src/composer/models";
import { composeQuery } from "@src/composer/query";
import { getTypedLiteralValue, refKeyFromRef } from "@src/composer/utils";
import {
  ActionDef,
  ActionHookDef,
  ChangesetDef,
  ChangesetOperationDef,
  CreateOneAction,
  DeleteOneAction,
  ExecuteHookAction,
  FetchOneAction,
  FieldSetter,
  FunctionName,
  QueryDef,
  UpdateOneAction,
} from "@src/types/definition";
import * as Spec from "@src/types/specification";

/**
 * Composes the custom actions block for an endpoint. Adds a default action
 * based on `endpoint.kind` if one is not defined in blueprint.
 * Requires `targets` to construct an initial variable context.
 */
export function composeActionBlock(specs: Spec.Action[]): ActionDef[] {
  // Collect actions from the spec, updating the context during the pass through.
  const actions = specs.map((atom) => {
    switch (atom.kind) {
      case "create":
      case "update": {
        return composeModelAction(atom);
      }
      case "delete": {
        return composeDeleteAction(atom);
      }
      case "execute": {
        return composeExecuteAction(atom);
      }
      case "fetch": {
        return composeFetchAction(atom);
      }
    }
  });

  return actions;
}

function composeDeleteAction(spec: FilteredByKind<Spec.Action, "delete">): DeleteOneAction {
  return {
    kind: "delete-one",
    targetPath: spec.targetPath.map((i) => i.text),
    model: findChangesetModel(spec.targetPath),
  };
}

function composeFetchAction(spec: FilteredByKind<Spec.Action, "fetch">): FetchOneAction {
  const changeset = spec.atoms.map((atom) => atomToChangesetOperation(atom, [], []));
  // fetch action's model is derived from it's query
  const query = composeQuery(spec.query);
  return {
    kind: "fetch-one",
    alias: spec.alias,
    changeset,
    model: query.retType,
    query: query,
  };
}

function composeExecuteAction(spec: FilteredByKind<Spec.Action, "execute">): ExecuteHookAction {
  const changeset = spec.atoms.map((atom) => atomToChangesetOperation(atom, [], []));

  const actionHook: ActionHookDef = {
    args: spec.hook.args.map((arg) => ({
      name: arg.name,
      setter: setterToFieldSetter(arg, changeset),
    })),
    hook: spec.hook.code,
  };

  return {
    kind: "execute-hook",
    hook: actionHook,
    changeset,
    responds: spec.responds,
  };
}

/**
 * Composes a single `ActionDef` based on current variable context, entrypoint, endpoint and action specs.
 */
function composeModelAction(spec: Spec.ModelAction): CreateOneAction | UpdateOneAction {
  const model = findChangesetModel(spec.targetPath);

  // TODO: this should be managed in resolver since it can be a circular reference which must be a compiler error
  const changeset: ChangesetDef = [];
  const resolveResult = resolveItems(
    // atoms to be resolved
    spec.actionAtoms,
    // item name resolver
    (atom) => {
      switch (atom.kind) {
        case "input":
          return atom.target.text;
        case "reference":
          return atom.target.text;
        case "set":
          return atom.target.text;
        case "virtual-input":
          return atom.name;
      }
    },
    // item resolver
    (atom) => {
      const op = atomToChangesetOperation(atom, spec.isPrimary ? [] : [spec.alias], changeset);
      // Add the changeset operation only if not added before
      if (!_.find(changeset, { name: op.name })) {
        changeset.push(op);
      }
    }
  );
  // handle error
  if (resolveResult.kind === "error") {
    console.log(
      "ERRORS",
      resolveResult.errors.map((e) => `${e.name} [${e.error.message ?? e.error}]`)
    );

    throw new Error(
      `Couldn't resolve all field setters: ${resolveResult.errors.map((i) => i.name).join()}`
    );
  }

  // Build the desired `ActionDef`.
  return modelActionFromParts(spec, model, changeset);
}

function expandSetterExpression(expr: Spec.Expr, changeset: ChangesetDef): FieldSetter {
  switch (expr.kind) {
    case "literal": {
      return getTypedLiteralValue(expr.literal);
    }
    case "identifier": {
      const [head, ...tail] = expr.identifier;
      const access = tail.map((i) => i.text);
      switch (head.ref.kind) {
        case "auth":
        case "model":
        case "queryTarget":
        case "target":
        case "action":
        case "repeat":
        case "struct": {
          return {
            kind: "reference-value",
            target: {
              alias: head.text,
              access,
            },
          };
        }
        case "authToken": {
          return {
            kind: "request-auth-token",
            access: ["user", "token"],
          };
        }
        case "virtualInput":
        case "modelAtom": {
          // if path has more than 1 element, it can't be a sibling call
          ensureEqual(access.length, 0, `Unexpected nested sibling ${head.text}: ${access}`);
          // check if sibling name is defined in the changeset
          const siblingOp = _.find(changeset, { name: head.text });
          if (siblingOp) {
            return { kind: "changeset-reference", referenceName: head.text };
          } else {
            throw new Error(`Circular reference: ${head.text}`);
          }
        }
        default:
          return assertUnreachable(head.ref);
      }
    }
    case "function": {
      return {
        kind: "function",
        name: expr.name as FunctionName, // FIXME proper validation
        args: expr.args.map((a) => expandSetterExpression(a, changeset)),
      };
    }
  }
}

function setterToChangesetOperation(
  atom: Spec.ActionAtomSet,
  changeset: ChangesetDef
): ChangesetOperationDef {
  return { name: atom.target.text, setter: setterToFieldSetter(atom.set, changeset) };
}

function setterToFieldSetter(
  set: Spec.ActionAtomSetHook | Spec.ActionAtomSetExp | Spec.ActionAtomSetQuery,
  changeset: ChangesetDef
): FieldSetter {
  switch (set.kind) {
    case "hook": {
      const args = set.hook.args.map((arg) => {
        const setter = setterToFieldSetter(arg, changeset);
        return { name: arg.name, setter };
      });
      return { kind: "fieldset-hook", hook: set.hook.code, args };
    }
    case "expression": {
      const exp = set.expr;
      return expandSetterExpression(exp, changeset);
    }
    case "query": {
      return { kind: "query", query: queryFromSpec(set.query) };
    }
  }
}

function atomToChangesetOperation(
  atom: Spec.ModelActionAtom,
  fieldsetNamespace: string[],
  changeset: ChangesetDef
): ChangesetOperationDef {
  switch (atom.kind) {
    case "virtual-input": {
      return {
        name: atom.name,
        setter: {
          kind: "fieldset-virtual-input",
          type: validateFieldType(atom.type),
          required: !atom.optional,
          nullable: atom.nullable,
          fieldsetAccess: [...fieldsetNamespace, atom.name],
          validators: composeValidators(validateFieldType(atom.type), atom.validators),
        },
      };
    }
    case "input": {
      const astType =
        atom.target.type.kind === "nullable" ? atom.target.type.type : atom.target.type;
      ensureEqual(astType.kind, "primitive");
      const type =
        astType.primitiveKind === "string"
          ? "text"
          : astType.primitiveKind === "float"
          ? "integer"
          : astType.primitiveKind;
      return {
        name: atom.target.text,
        setter: {
          kind: "fieldset-input",
          type,
          required: !atom.optional,
          fieldsetAccess: [...fieldsetNamespace, atom.target.text],
        },
      };
    }
    case "reference": {
      return {
        name: atom.target.text,
        setter: {
          kind: "fieldset-reference-input",
          throughRefKey: refKeyFromRef(atom.through.ref),
          fieldsetAccess: [...fieldsetNamespace, `${atom.target.text}_${atom.through.text}`],
        },
      };
    }
    case "set": {
      return setterToChangesetOperation(atom, changeset);
    }
  }
}

/**
 * Returns a model the changeset operates on. Taken from the end of the resolved path
 * which must not end with a `leaf`.
 *
 * FIXME this function is not specific to `changeset`, rename. This may be deprecated
 *       by proposed changes in `getTypedPathFromContext`.
 */
function findChangesetModel(specTargetPath: Spec.IdentifierRef[]): string {
  const modelName = getTypeModel(specTargetPath.at(-1)!.type);
  if (!modelName) return getTypeModel(specTargetPath.at(-2)!.type)!;
  return modelName;
}

/**
 * Constructs an `ActionDef` for a model action.
 */
function modelActionFromParts(
  spec: Spec.ModelAction,
  model: string,
  changeset: ChangesetDef
): CreateOneAction | UpdateOneAction {
  switch (spec.kind) {
    case "create": {
      return {
        kind: "create-one",
        alias: spec.alias,
        changeset,
        targetPath: spec.targetPath.map((i) => i.text),
        model,
        select: [],
        isPrimary: spec.isPrimary,
      };
    }
    case "update": {
      // FIXME update-many when targetKind is model
      return {
        kind: "update-one",
        changeset,
        alias: spec.alias,
        targetPath: spec.targetPath.map((i) => i.text),
        model,
        filter: undefined,
        select: [],
        isPrimary: spec.isPrimary,
      };
    }
  }
}

export function queryFromSpec(qspec: Spec.Query): QueryDef {
  ensureEmpty(qspec.aggregate, "Aggregates are not yet supported in action queries");

  const pathPrefix = _.first(qspec.from);
  ensureExists(pathPrefix, `Action query "fromModel" path is empty ${qspec.from}`);

  return composeQuery(qspec);
}
