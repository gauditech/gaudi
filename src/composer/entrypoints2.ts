import { kindFilter } from "@src/common/kindFilter";
import * as AST from "@src/compiler/ast/ast";
import { getTypeModel } from "@src/compiler/ast/type";
import { SelectDef, SelectItem } from "@src/types/definition";

export function processSelect(
  models: AST.Model[],
  model: string,
  select: AST.Select | undefined,
  parentNamePath: string[]
): SelectDef {
  if (select === undefined) {
    const modelAtoms = models.find((m) => m.name.text === model)?.atoms ?? [];
    const idField: SelectItem = {
      kind: "field",
      name: "id",
      alias: "id",
      namePath: [...parentNamePath, "id"],
      refKey: `${model}.id`,
    };
    const fields = kindFilter(modelAtoms, "field").map(
      (f): SelectItem => ({
        kind: "field",
        name: f.name.text,
        alias: f.name.text,
        namePath: [...parentNamePath, f.name.text],
        refKey: `${model}.${f.name.text}`,
      })
    );
    const references = kindFilter(modelAtoms, "reference").map(
      (r): SelectItem => ({
        kind: "field",
        name: `${r.name.text}_id`,
        alias: `${r.name.text}_id`,
        namePath: [...parentNamePath, `${r.name.text}_id`],
        refKey: `${model}.${r.name.text}_id`,
      })
    );
    return [idField, ...fields, ...references];
  }

  return select.map(({ target, select }): SelectItem => {
    if (target.kind === "long") {
      throw Error("Long select form unsupported composer");
    }
    if (target.name.ref.kind !== "modelAtom") {
      throw new Error("Unexpected select ref");
    }

    const namePath = [...parentNamePath, target.name.ref.name];

    switch (target.name.ref.atomKind) {
      case "field":
      case "computed":
        return {
          kind: target.name.ref.atomKind,
          refKey: `${target.name.ref.model}.${target.name.ref.name}`,
          name: target.name.ref.name,
          alias: target.name.identifier.text,
          namePath,
        };
      case "hook":
        return {
          kind: "model-hook",
          refKey: `${target.name.ref.model}.${target.name.ref.name}`,
          name: target.name.ref.name,
          alias: target.name.identifier.text,
          namePath,
        };
      case "reference":
      case "relation":
      case "query": {
        const targetModel = getTypeModel(target.name.type);
        // if type has no model, it's an aggregate
        if (!targetModel) {
          return {
            kind: "aggregate",
            refKey: `${target.name.ref.model}.${target.name.ref.name}`,
            name: target.name.ref.name,
            alias: target.name.identifier.text,
            namePath,
          };
        }
        return {
          kind: target.name.ref.atomKind,
          name: target.name.ref.name,
          alias: target.name.identifier.text,
          namePath,
          select: processSelect(models, targetModel, select, namePath),
        };
      }
    }
  });
}
