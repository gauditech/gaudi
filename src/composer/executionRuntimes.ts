import { ensureUnique } from "@src/common/utils";
import { Definition, ExecutionRuntimeDef } from "@src/types/definition";
import { ExecutionRuntimeSpec } from "@src/types/specification";

const EXECUTION_RUNTIME_GAUDI_INTERNAL = "$GAUDI_INTERNAL";

export function composeExecutionRuntimes(def: Definition, runtimes: ExecutionRuntimeSpec[]): void {
  def.runtimes = runtimes.map((p) => composeRuntime(def, p));

  // there must be one and only one default runtime
  // do this BEFORE injecting internal runtime
  processDefaultRuntime(def);

  // inject gaudi internal runtime
  def.runtimes.push(composeInternalExecutionRuntime(def));

  // check for duplicate names
  ensureUnique(
    def.runtimes.map((r) => r.name.toLowerCase()),
    "Execution runtime names must be unique"
  );
}

export function getInternalExecutionRuntimeName(): string {
  return EXECUTION_RUNTIME_GAUDI_INTERNAL;
}

function composeInternalExecutionRuntime(def: Definition): ExecutionRuntimeDef {
  return composeRuntime(def, {
    name: getInternalExecutionRuntimeName(),
    default: false,
    sourcePath: "./hook",
  });
}

function composeRuntime(def: Definition, runtime: ExecutionRuntimeSpec): ExecutionRuntimeDef {
  const name = runtime.name;
  const sourcePath = runtime.sourcePath;

  return {
    name,
    // only "node" type is currently available
    type: "node",
    default: !!runtime.default,
    sourcePath,
  };
}

/**
 * Make sure that there is one and only one default runtime.
 * Throw error otherwise.
 */
function processDefaultRuntime(def: Definition): void {
  const runtimeCount = def.runtimes.length;
  const defaultRuntimeCount = def.runtimes.filter((r) => r.default).length;

  // exactly 1 default - OK
  if (defaultRuntimeCount === 1) return;

  // exactly 0 defaults
  if (defaultRuntimeCount === 0) {
    // no runtimes at all - OK
    if (runtimeCount === 0) return;

    // if it's a single runtime - make it default by default
    if (runtimeCount === 1) {
      // update execution runtime's default prop
      def.runtimes[0].default = true;
      return;
    }
  }

  throw new Error("There can be only one default execution runtime");
}
