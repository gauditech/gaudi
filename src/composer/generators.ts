import { assertUnreachable } from "@src/common/utils";
import { Definition, GeneratorDef } from "@src/types/definition";
import { Generator } from "@src/types/specification";

export function composeGenerators(def: Definition, generators: Generator[]): void {
  def.generators = generators.map((g) => composeGenerator(def, g));
}

function composeGenerator(def: Definition, generator: Generator): GeneratorDef {
  const kind = generator.kind;
  switch (kind) {
    case "generator-client": {
      // target
      const target = generator.target;
      if (target !== "js") {
        throw new Error(`Unsupported client generator target "${target}"`);
      }

      // api
      const api = generator.api;
      if (api !== "entrypoint" && api !== "model") {
        throw new Error(`Unsupported client generator api "${api}"`);
      }

      // output
      const output = generator.output;

      return {
        kind: "generator-client",
        target,
        api,
        output,
      };
    }
    default:
      assertUnreachable(kind);
  }
}
