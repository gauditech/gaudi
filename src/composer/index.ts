import fs from "fs";
import Arcade from "../examples/arcade/definition";
import { setupDb } from "./dbSetup";
import { readDefinition } from "./parser";
const definition = readDefinition(Arcade.models);
// ok now we have definition, let's build something with it!
fs.writeFileSync(
  "examples/arcade/_data.json",
  JSON.stringify(definition, null, 2)
);

const dbSetup = setupDb(definition);
fs.writeFileSync("examples/arcade/_db_setup.sql", dbSetup);
