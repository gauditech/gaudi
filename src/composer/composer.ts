import { composeEntrypoints } from "./entrypoints";
import { composeModels } from "./models";

import { composePopulators } from "@src/composer/populators";
import { Definition } from "@src/types/definition";
import { Specification } from "@src/types/specification";

export function compose(input: Specification): Definition {
  // let's start with empty definition
  // sub-composers are expected to mutate it
  const def: Definition = { models: [], entrypoints: [], resolveOrder: [], populators: [] };
  composeModels(def, input.models);
  composeEntrypoints(def, input.entrypoints);
  composePopulators(def, input.populators);

  return def;
}
