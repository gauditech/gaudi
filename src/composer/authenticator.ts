import _ from "lodash";

import { getRef } from "@src/common/refs";
import { assertUnreachable } from "@src/common/utils";
import { compile } from "@src/compiler/compiler";
import { getInternalExecutionRuntimeName } from "@src/composer/executionRuntimes";
import { parse } from "@src/parser/parser";
import {
  AuthenticatorMethodDef,
  AuthenticatorNamedModelDef,
  Definition,
} from "@src/types/definition";
import {
  AuthenticatorMethodSpec,
  AuthenticatorSpec,
  Specification,
} from "@src/types/specification";

/**
 * Compose authenticator block.
 */
export function composeAuthenticator(def: Definition, spec: AuthenticatorSpec | undefined): void {
  if (spec == undefined) {
    return;
  }

  // hardcoded authenticator name - not exposed through blueprint cause we don't support multiple auth blocks yet
  const name = "Auth";
  const authUserModel = composeTargetModel(def, spec.authUserModelName);
  const accessTokenModel = composeTargetModel(def, spec.accessTokenModelName);
  const method = composeMethod(def, spec.method);

  def.authenticator = {
    name,
    authUserModel,
    accessTokenModel,
    method,
  };
}

function composeTargetModel(def: Definition, modelName: string): AuthenticatorNamedModelDef {
  // get authenticator target model (injected in compiler)
  const model = getRef.model(def, modelName);

  return {
    name: modelName,
    refKey: model.refKey,
  };
}

function composeMethod(
  def: Definition,
  methodSpec: AuthenticatorMethodSpec
): AuthenticatorMethodDef {
  const kind = methodSpec.kind;
  if (kind === "basic") {
    return {
      kind,
    };
  } else {
    assertUnreachable(kind);
  }
}

/**
 * Create authenticator specs.
 */
export function compileAuthenticatorSpec(
  authenticatorSpec: AuthenticatorSpec | undefined
): Specification {
  if (authenticatorSpec == null) {
    return { models: [], entrypoints: [], populators: [], runtimes: [], authenticator: undefined };
  }

  const authUserModelName = authenticatorSpec.authUserModelName;
  const accessTokenModelName = authenticatorSpec.accessTokenModelName;

  const internalExecRuntimeName = getInternalExecutionRuntimeName();

  const AuthenticatorModel = `
    model ${authUserModelName} {
      field name { type text, validate { min 1 } }
      field username { type text, unique, validate { min 8 } }
      field passwordHash { type text, validate { min 8 } }
      relation tokens { from ${accessTokenModelName}, through authUser }
    }
    model ${accessTokenModelName} {
      field token { type text, unique }
      field expiryDate { type text }
      reference authUser { to ${authUserModelName} }
    }

    entrypoint Auth {
      target model ${authUserModelName} as authUser

      // login
      custom endpoint {
        path "login"
        method POST
        cardinality many

        action {
          fetch as existingAuthUser {
            virtual input username { type text }

            query {
              from ${authUserModelName} as a
              filter a.username is username
              limit 1
            }
            // TODO: throw error id user is not resolved
            // currently, "existingAuthUser" ends up empty and "authenticateUser" hook throws error
          }

          execute {
            virtual input password { type text }

            hook {
              arg clearPassword password
              arg hashPassword existingAuthUser.passwordHash
    
              runtime ${internalExecRuntimeName}
              source authenticateUser from "hooks/actions.js"
            }
          }

          // create access token
          create ${accessTokenModelName} as accessToken {
            set token cryptoToken(32)
            set expiryDate stringify(now() + 3600000) // 1 hour
            set authUser existingAuthUser
          }

          // return token
          execute {
            responds

            hook {
              arg status 200
              arg body_token accessToken.token

              runtime ${internalExecRuntimeName}
              source sendResponse from "hooks/actions.js"
            }
          }
        }
      }

      // logout
      custom endpoint {
        path "logout"
        method POST
        cardinality many

        action {
          fetch ${accessTokenModelName} as accessToken {
 
            query {
              from ${accessTokenModelName}
              filter token is @requestAuthToken
              limit 1
            }
          }

          delete accessToken {}

          // return status
          execute {
            responds

            hook {
              arg status 204

              runtime ${internalExecRuntimeName}
              source sendResponse from "hooks/actions.js"
            }
          }
        }
      }

      // register
      custom endpoint {
        path "register"
        method POST
        cardinality many

        action {
          create authUser {
            input { name, username }
            virtual input password { type text, validate { min 8 } }
            set passwordHash cryptoHash(password, 10)
          }

          // return created user
          execute {
            responds

            hook {
              arg status 201
              // body - see hook for details
              arg body_id authUser.id
              arg body_name authUser.name
              arg body_username authUser.username

              runtime ${internalExecRuntimeName}
              source sendResponse from "hooks/actions.js"
            }
          }
        }
      }
    }
  `;

  return compile(parse(AuthenticatorModel));
}
