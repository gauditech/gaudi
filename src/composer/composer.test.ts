import { compose } from "./composer";

import { compile } from "@src/compiler/compiler";
import definitionInput from "@src/composer/tests/data/definition.json";
import specificationInput from "@src/composer/tests/data/specification.json";
import { parse } from "@src/parser/parser";
import { Specification } from "@src/types/specification";

describe("compose models", () => {
  it("doesn't crash on empty blueprint", () => {
    expect(() => compose(compile(parse("")))).not.toThrow();
  });
  it("correctly composes the git example", () => {
    expect(compose(specificationInput)).toStrictEqual(definitionInput);
  });

  it("fails on case insensitive duplicate field name", () => {
    const specification: Specification = {
      models: [
        {
          name: "Org",
          fields: [
            { name: "name", type: "text" },
            { name: "Name", type: "text" },
          ],
          relations: [],
          references: [],
          queries: [],
          computeds: [],
        },
      ],
      entrypoints: [],
    };
    expect(() => compose(specification)).toThrowError("Items not unique!");
  });
  it("fails on name colision between field and reference", () => {
    const specification: Specification = {
      models: [
        {
          name: "Org",
          fields: [{ name: "name", type: "text" }],
          references: [{ name: "name", toModel: "Org" }],
          relations: [],
          queries: [],
          computeds: [],
        },
      ],
      entrypoints: [],
    };
    expect(() => compose(specification)).toThrowError("Items not unique!");
  });
  it("fails when relation doesn't point to a reference", () => {
    const specification: Specification = {
      models: [
        {
          name: "Org",
          fields: [{ name: "name", type: "text" }],
          references: [],
          relations: [{ name: "repos", fromModel: "Repo", through: "name" }],
          queries: [],
          computeds: [],
        },
        {
          name: "Repo",
          fields: [{ name: "name", type: "text" }],
          references: [{ name: "org", toModel: "Org" }],
          relations: [],
          queries: [],
          computeds: [],
        },
      ],
      entrypoints: [],
    };
    expect(() => compose(specification)).toThrowError(
      "Expecting type reference but found a type field"
    );
  });
  it("detect infinite loop when resolving", () => {
    const bp = `
    model Org { reference no { to Unknown } }
    `;
    expect(() => compose(compile(parse(bp)))).toThrowError("infinite-loop");
  });
  it("parses validators", () => {
    const bp = `
    model Org {
      field adminEmail { type text, validate { min 4, max 100, isEmail } }
      field num_employees { type integer, validate { min 0, max 9999 } }
    }`;
    const def = compose(compile(parse(bp)));
    expect(def).toMatchSnapshot();
  });
  it("fails on invalid validator", () => {
    const bp = `
    model Org {
      field adminEmail { type text }
      field num_employees { type integer, validate { isEmail } }
    }`;
    const spec = compile(parse(bp));
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(`"Unknown validator!"`);
  });
});
