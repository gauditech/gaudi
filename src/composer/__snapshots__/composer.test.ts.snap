// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose models parses model references 1`] = `
[
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "parentitem",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "ParentItem",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "ParentItem.id",
        "type": "integer",
        "unique": true,
        "validators": [],
      },
      {
        "dbname": "itemnoaction_id",
        "kind": "field",
        "modelRefKey": "ParentItem",
        "name": "itemNoAction_id",
        "nullable": false,
        "primary": false,
        "refKey": "ParentItem.itemNoAction_id",
        "type": "integer",
        "unique": false,
        "validators": [],
      },
      {
        "dbname": "itemcascade_id",
        "kind": "field",
        "modelRefKey": "ParentItem",
        "name": "itemCascade_id",
        "nullable": false,
        "primary": false,
        "refKey": "ParentItem.itemCascade_id",
        "type": "integer",
        "unique": false,
        "validators": [],
      },
      {
        "dbname": "itemsetnull_id",
        "kind": "field",
        "modelRefKey": "ParentItem",
        "name": "itemSetNull_id",
        "nullable": true,
        "primary": false,
        "refKey": "ParentItem.itemSetNull_id",
        "type": "integer",
        "unique": false,
        "validators": [],
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "ParentItem",
    "queries": [],
    "refKey": "ParentItem",
    "references": [
      {
        "fieldRefKey": "ParentItem.itemNoAction_id",
        "kind": "reference",
        "modelRefKey": "ParentItem",
        "name": "itemNoAction",
        "nullable": false,
        "onDelete": undefined,
        "refKey": "ParentItem.itemNoAction",
        "toModelFieldRefKey": "ReferencedItem1.id",
        "toModelRefKey": "ReferencedItem1",
        "unique": false,
      },
      {
        "fieldRefKey": "ParentItem.itemCascade_id",
        "kind": "reference",
        "modelRefKey": "ParentItem",
        "name": "itemCascade",
        "nullable": false,
        "onDelete": "cascade",
        "refKey": "ParentItem.itemCascade",
        "toModelFieldRefKey": "ReferencedItem2.id",
        "toModelRefKey": "ReferencedItem2",
        "unique": false,
      },
      {
        "fieldRefKey": "ParentItem.itemSetNull_id",
        "kind": "reference",
        "modelRefKey": "ParentItem",
        "name": "itemSetNull",
        "nullable": true,
        "onDelete": "setNull",
        "refKey": "ParentItem.itemSetNull",
        "toModelFieldRefKey": "ReferencedItem3.id",
        "toModelRefKey": "ReferencedItem3",
        "unique": false,
      },
    ],
    "relations": [],
  },
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "referenceditem1",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "ReferencedItem1",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "ReferencedItem1.id",
        "type": "integer",
        "unique": true,
        "validators": [],
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "ReferencedItem1",
    "queries": [],
    "refKey": "ReferencedItem1",
    "references": [],
    "relations": [
      {
        "fromModel": "ParentItem",
        "fromModelRefKey": "ParentItem",
        "kind": "relation",
        "modelRefKey": "ReferencedItem1",
        "name": "parent",
        "refKey": "ReferencedItem1.parent",
        "through": "itemNoAction",
        "throughRefKey": "ParentItem.itemNoAction",
        "unique": false,
      },
    ],
  },
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "referenceditem2",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "ReferencedItem2",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "ReferencedItem2.id",
        "type": "integer",
        "unique": true,
        "validators": [],
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "ReferencedItem2",
    "queries": [],
    "refKey": "ReferencedItem2",
    "references": [],
    "relations": [
      {
        "fromModel": "ParentItem",
        "fromModelRefKey": "ParentItem",
        "kind": "relation",
        "modelRefKey": "ReferencedItem2",
        "name": "parent",
        "refKey": "ReferencedItem2.parent",
        "through": "itemCascade",
        "throughRefKey": "ParentItem.itemCascade",
        "unique": false,
      },
    ],
  },
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "referenceditem3",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "ReferencedItem3",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "ReferencedItem3.id",
        "type": "integer",
        "unique": true,
        "validators": [],
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "ReferencedItem3",
    "queries": [],
    "refKey": "ReferencedItem3",
    "references": [],
    "relations": [
      {
        "fromModel": "ParentItem",
        "fromModelRefKey": "ParentItem",
        "kind": "relation",
        "modelRefKey": "ReferencedItem3",
        "name": "parent",
        "refKey": "ReferencedItem3.parent",
        "through": "itemSetNull",
        "throughRefKey": "ParentItem.itemSetNull",
        "unique": false,
      },
    ],
  },
]
`;

exports[`compose models parses validators 1`] = `
[
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "org",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "Org",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "Org.id",
        "type": "integer",
        "unique": true,
        "validators": [],
      },
      {
        "dbname": "adminemail",
        "kind": "field",
        "modelRefKey": "Org",
        "name": "adminEmail",
        "nullable": false,
        "primary": false,
        "refKey": "Org.adminEmail",
        "type": "string",
        "unique": false,
        "validators": [
          {
            "args": [
              {
                "kind": "integer",
                "value": 4,
              },
            ],
            "inputType": "string",
            "name": "minLength",
          },
          {
            "args": [
              {
                "kind": "integer",
                "value": 100,
              },
            ],
            "inputType": "string",
            "name": "maxLength",
          },
          {
            "args": [],
            "inputType": "string",
            "name": "isEmail",
          },
        ],
      },
      {
        "dbname": "num_employees",
        "kind": "field",
        "modelRefKey": "Org",
        "name": "num_employees",
        "nullable": false,
        "primary": false,
        "refKey": "Org.num_employees",
        "type": "integer",
        "unique": false,
        "validators": [
          {
            "args": [
              {
                "kind": "integer",
                "value": 0,
              },
            ],
            "inputType": "integer",
            "name": "minInt",
          },
          {
            "args": [
              {
                "kind": "integer",
                "value": 9999,
              },
            ],
            "inputType": "integer",
            "name": "maxInt",
          },
        ],
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "Org",
    "queries": [],
    "refKey": "Org",
    "references": [],
    "relations": [],
  },
]
`;
